Assignment 2

ADi2991 opened the pull request:
Sorry for the late submission. Won't happen again.
Aditya Rao (47221503)

ADi2991 committed:
First commit, start of development.

ADi2991 committed:
Dependencies added, struct initialized.

ADi2991 committed:
New game implemented with test

ADi2991 committed:
Tally implemented with test

ADi2991 committed:
3 tests left

ADi2991 committed:
Final version, game and tests run successfully.

ADi2991 committed:
Rebase to branch Assignment-2

pragdave commented on hangman/lib/game/game.ex:
```diff
@@ -0,0 +1,87 @@
+defmodule Hangman.Game do
+  def new_game() do 
+    # function to filter alphabets
+    alphabets_only = fn x -> if Regex.match?(~r{[a-zA-Z]}, x),
```
What's the difference between

~~~ elixir
if Regex.match?(~r{[a-zA-Z]}, x),
 do: true,
 else: false
~~~

and

~~~ elixir
Regex.match?(~r{[a-zA-Z]}, x)
~~~

pragdave commented on hangman/lib/game/game.ex:
```diff
@@ -0,0 +1,87 @@
+  end
+
+  def tally(game) do
+    game
```
This is a dangerous technique. What happens if you added a couple of secret fields to the struct. If you then forgot to update this code, you'd end up passing them to the client.

pragdave commented on hangman/lib/game/game.ex:
```diff
@@ -0,0 +1,87 @@
+      make_move(game, guess, :new_guess)
+    end
+  end
+  def make_move(game, guess, _new_guess) do
```
If I were grading this, the size and complexity of this function would lose you marks.

Try to keep your functions to just a couple of lines, and think of ways of avoiding all the `if` statements. It takes a little more thinking up front, but results in code that "just works".

The trick it to ask yourself "what one thing does this function do?" Write that, in terms of other functions. Then, for each of those, ask "what does this do."

pragdave commented:
This breaks my heart—you seem to have a working application, but you were an hour late.

I'm going to break my own rules, just this once.

If you go through this code and tidy up the functions in Game to make them more functional, check it in and send me an email, I'll grade it, with a 10 point penalty. I'll need to see it by Sunday, 23:59.

I know you can make this tidier. Show me.

ADi2991 commented:
Thank you so much for your faith in me. I will do my best. I have one small
request though:
I know my code quality is far from perfect. I wish to learn from your
experience. Kindly grade my assignments as tough as you can. I could use
the nitpicking.

On Tue, Oct 3, 2017, 11:51 Dave Thoms <notifications@github.com> wrote:

> This breaks my heart—you seem to have a working application, but you were
> an hour late.
>
> I'm going to break my own rules, just this once.
>
> If you go through this code and tidy up the functions in Game to make them
> more functional, check it in and send me an email, I'll grade it, with a 10
> point penalty. I'll need to see it by Sunday, 23:59.
>
> I know you can make this tidier. Show me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cse-x345-fall-2017/2-first-hangman/pull/19#issuecomment-333907956>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKRoTm5ipy6YZsiycK2qmelEkj1QtpKCks5somX2gaJpZM4PqcNC>
> .
>

ADi2991 committed:
Evaluate function works correctly (verified by test)

Commit:
1. 'evaluate' changed to 'fill_occurences'
2. Introduced separate test for helper

ADi2991 committed:
1. Helper on_good_guess introduced
2. Test for on_good_guess introduced and runs successfully

ADi2991 committed:
1. Introduced helper on_bad_guess
2. Introduced test for the helper, runs successfully

ADi2991 committed:
1. Introduced function give_feedback
2. Introduced test(s) for give_feedback which run well

ADi2991 committed:
1. Changes made to some method signatures to allow pipelining, tests updated to work.
2. :already_guessed detection functionality added.
3. result of give_feedback() is now sent using send_updated _state() instead of directly through make_move()
4. tally() now includes last_guess, and the arguments are now more concise
5. redundancy in fill_occurences() removed and comments refined.

ADi2991 committed:
Merge branch 'master' into Assignment-2

ADi2991 commented:
I have tried my best and revised the code quite a bit. Hope you like it.
Thanks again :)
